File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/creating-jvm-language-7cdaf40c49/job_byox_build_v1_5d3af01bcd55594749205d85322b9429/attempt-001/sealed/reference/src/main/java/dev/learningfactory/sprig/Lexer.java
package dev.learningfactory.sprig;

import java.util.ArrayList;
import java.util.List;

final class Lexer {
    private static final int MAX_TOKENS = 100_000;

    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int index;
    private int line = 1;
    private int column = 1;

    Lexer(String source) {
        this.source = source;
    }

    List<Token> scan() throws CompileFailure {
        while (!atEnd()) {
            skipTrivia();
            if (atEnd()) break;
            scanToken();
        }
        tokens.add(new Token(TokenType.EOF, "", line, column));
        return List.copyOf(tokens);
    }

    private void skipTrivia() {
        boolean again;
        do {
            again = false;
            while (!atEnd()) {
                char c = peek();
                if (c == ' ' || c == '\t' || c == '\n' || c == '\r') {
                    advance();
                } else {
                    break;
                }
            }
            if (!atEnd() && peek() == '/' && peekNext() == '/') {
                advance();
                advance();
                while (!atEnd() && peek() != '\n' && peek() != '\r') advance();
                again = true;
            }
        } while (again);
    }

    private void scanToken() throws CompileFailure {
        int start = index;
        int startLine = line;
        int startColumn = column;
        char c = advance();
        if (isIdentifierStart(c)) {
            while (isIdentifierPart(peek())) advance();
            String text = source.substring(start, index);
            add(keyword(text), text, startLine, startColumn);
            return;
        }
        if (isDigit(c)) {
            if (c == '0') {
                add(TokenType.INTEGER, "0", startLine, startColumn);
                return;
            }
            long value = c - '0';
            while (isDigit(peek())) {
                value = value * 10 + (advance() - '0');
                if (value > Integer.MAX_VALUE) {
                    while (isDigit(peek())) advance();
                    throw CompileFailure.at("E_INT_RANGE", startLine, startColumn,
                            "integer literal exceeds 2147483647");
                }
            }
            add(TokenType.INTEGER, source.substring(start, index), startLine, startColumn);
            return;
        }

        TokenType type = switch (c) {
            case '(' -> TokenType.LEFT_PAREN;
            case ')' -> TokenType.RIGHT_PAREN;
            case '{' -> TokenType.LEFT_BRACE;
            case '}' -> TokenType.RIGHT_BRACE;
            case ';' -> TokenType.SEMICOLON;
            case '+' -> TokenType.PLUS;
            case '*' -> TokenType.STAR;
            case '/' -> TokenType.SLASH;
            case '%' -> TokenType.PERCENT;
            case '-' -> match('>') ? TokenType.ARROW : TokenType.MINUS;
            case '=' -> match('=') ? TokenType.EQUAL_EQUAL : TokenType.ASSIGN;
            case '!' -> match('=') ? TokenType.BANG_EQUAL : TokenType.BANG;
            case '<' -> match('=') ? TokenType.LESS_EQUAL : TokenType.LESS;
            case '>' -> match('=') ? TokenType.GREATER_EQUAL : TokenType.GREATER;
            case '&' -> {
                if (!match('&')) {
                    throw CompileFailure.at("E_CHAR", startLine, startColumn,
                            "a lone '&' is not an operator");
                }
                yield TokenType.AND_AND;
            }
            case '|' -> {
                if (!match('|')) {
                    throw CompileFailure.at("E_CHAR", startLine, startColumn,
                            "a lone '|' is not an operator");
                }
                yield TokenType.OR_OR;
            }
            default -> throw CompileFailure.at("E_CHAR", startLine, startColumn,
                    "unexpected character U+" + String.format("%04X", (int) c));
        };
        add(type, source.substring(start, index), startLine, startColumn);
    }

    private void add(TokenType type, String text, int tokenLine, int tokenColumn)
            throws CompileFailure {
        if (tokens.size() >= MAX_TOKENS) {
            throw CompileFailure.at("E_LIMIT", tokenLine, tokenColumn,
                    "token limit of " + MAX_TOKENS + " exceeded");
        }
        tokens.add(new Token(type, text, tokenLine, tokenColumn));
    }

    private static TokenType keyword(String text) {
        return switch (text) {
            case "fn" -> TokenType.FN;
            case "main" -> TokenType.MAIN;
            case "Int" -> TokenType.INT;
            case "Bool" -> TokenType.BOOL;
            case "let" -> TokenType.LET;
            case "print" -> TokenType.PRINT;
            case "if" -> TokenType.IF;
            case "else" -> TokenType.ELSE;
            case "while" -> TokenType.WHILE;
            case "return" -> TokenType.RETURN;
            case "true" -> TokenType.TRUE;
            case "false" -> TokenType.FALSE;
            default -> TokenType.IDENTIFIER;
        };
    }

    private boolean match(char expected) {
        if (atEnd() || peek() != expected) return false;
        advance();
        return true;
    }

    private char advance() {
        char c = source.charAt(index++);
        if (c == '\r') {
            if (!atEnd() && source.charAt(index) == '\n') index++;
            line++;
            column = 1;
        } else if (c == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        return c;
    }

    private char peek() {
        return atEnd() ? '\0' : source.charAt(index);
    }

    private char peekNext() {
        return index + 1 >= source.length() ? '\0' : source.charAt(index + 1);
    }

    private boolean atEnd() {
        return index >= source.length();
    }

    private static boolean isIdentifierStart(char c) {
        return c == '_' || c >= 'A' && c <= 'Z' || c >= 'a' && c <= 'z';
    }

    private static boolean isIdentifierPart(char c) {
        return isIdentifierStart(c) || isDigit(c);
    }

    private static boolean isDigit(char c) {
        return c >= '0' && c <= '9';
    }
}

File: learning-factory/warehouse/artifacts/projects/build-your-own-x/programming-language/a-compiler-from-scratch-84427e53fe/job_byox_build_v1_f3e0d99cf97b485202190e3be4a0e8f6/attempt-001/sealed/reference/lib/pebble/lexer.rb
# frozen_string_literal: true

module Pebble
  class Lexer
    KEYWORDS = {
      "let" => :LET,
      "print" => :PRINT,
      "if" => :IF,
      "else" => :ELSE,
      "while" => :WHILE,
      "true" => :TRUE,
      "false" => :FALSE
    }.freeze

    SINGLE_TOKENS = {
      "+" => :PLUS,
      "-" => :MINUS,
      "*" => :STAR,
      "%" => :PERCENT,
      "(" => :LEFT_PAREN,
      ")" => :RIGHT_PAREN,
      "{" => :LEFT_BRACE,
      "}" => :RIGHT_BRACE,
      ";" => :SEMICOLON
    }.freeze

    MAX_LITERAL = 2_147_483_647

    def initialize(source)
      unless source.is_a?(String)
        raise LexError.new("source must be a String", 1, 1)
      end

      @source = source
      @tokens = []
      @start = 0
      @current = 0
      @line = 1
      @column = 1
      @start_line = 1
      @start_column = 1
    end

    def scan_tokens
      until at_end?
        @start = @current
        @start_line = @line
        @start_column = @column
        scan_token
      end
      @tokens << Token.new(:EOF, "", nil, @line, @column)
      @tokens
    end

    private

    def scan_token
      character = advance
      single_type = SINGLE_TOKENS[character]
      if single_type
        add_token(single_type)
        return
      end

      case character
      when " ", "\t", "\r", "\n"
        return
      when "/"
        if match?("/")
          advance until at_end? || peek == "\n"
        else
          add_token(:SLASH)
        end
      when "!"
        add_token(match?("=") ? :BANG_EQUAL : :BANG)
      when "="
        add_token(match?("=") ? :EQUAL_EQUAL : :EQUAL)
      when "<"
        add_token(match?("=") ? :LESS_EQUAL : :LESS)
      when ">"
        add_token(match?("=") ? :GREATER_EQUAL : :GREATER)
      else
        if digit?(character)
          scan_integer(character)
        elsif identifier_start?(character)
          scan_identifier
        else
          raise LexError.new(
            "unexpected character #{character.inspect}",
            @start_line,
            @start_column
          )
        end
      end
    end

    def scan_integer(first_character)
      advance while digit?(peek)
      lexeme = current_lexeme
      if first_character == "0" && lexeme.length > 1
        raise LexError.new("leading zero in integer literal", @start_line, @start_column)
      end

      value = lexeme.to_i
      if value > MAX_LITERAL
        raise LexError.new("integer literal is out of range", @start_line, @start_column)
      end
      add_token(:INTEGER, value)
    end

    def scan_identifier
      advance while identifier_part?(peek)
      lexeme = current_lexeme
      type = KEYWORDS.fetch(lexeme, :IDENTIFIER)
      literal = if type == :TRUE
                  true
                elsif type == :FALSE
                  false
                end
      add_token(type, literal)
    end

    def advance
      character = @source[@current]
      @current += 1
      if character == "\n"
        @line += 1
        @column = 1
      else
        @column += 1
      end
      character
    end

    def match?(expected)
      return false if at_end? || @source[@current] != expected

      advance
      true
    end

    def peek
      at_end? ? "\0" : @source[@current]
    end

    def at_end?
      @current >= @source.length
    end

    def current_lexeme
      @source[@start...@current]
    end

    def add_token(type, literal = nil)
      @tokens << Token.new(type, current_lexeme, literal, @start_line, @start_column)
    end

    def digit?(character)
      character >= "0" && character <= "9"
    end

    def identifier_start?(character)
      (character >= "a" && character <= "z") ||
        (character >= "A" && character <= "Z") || character == "_"
    end

    def identifier_part?(character)
      identifier_start?(character) || digit?(character)
    end
  end
end
